ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java

```diff
@@ -740,16 +742,21 @@ static KsqlRestApplication buildApplication(
     return Optional.empty();
   }
 
-  private static RoutingFilters initializeRoutingFilters(
+  private static RoutingFilterFactory initializeRoutingFilterFactory(
       final KsqlConfig ksqlConfig,
-      final Optional<HeartbeatAgent> heartbeatAgent) {
-    final ImmutableList.Builder<RoutingFilter> filterBuilder = ImmutableList.builder();
-    if (!ksqlConfig.getBoolean(KsqlConfig.KSQL_QUERY_PULL_ENABLE_STANDBY_READS)) {
-      filterBuilder.add(new ActiveHostFilter());
-    }
-    filterBuilder.add(new LivenessFilter(heartbeatAgent));
-    final RoutingFilters routingFilters = new RoutingFilters(filterBuilder.build());
-    return routingFilters;
+      final Optional<HeartbeatAgent> heartbeatAgent,
+      final Optional<LagReportingAgent> lagReportingAgent) {
+    return (routingOptions, hosts, active, applicationQueryId, storeName, partition) -> {
+      final ImmutableList.Builder<RoutingFilter> filterBuilder = ImmutableList.builder();
+      if (!ksqlConfig.getBoolean(KsqlConfig.KSQL_QUERY_PULL_ENABLE_STANDBY_READS)) {
+        filterBuilder.add(new ActiveHostFilter(active));
+      }
```

Comment:
for my understanding, if standby reads are disabled why do we bother adding the other filters?

Reply:
You could argue that there's value in allowing for these filters, even if standby reads are disabled. They would allow for failing the request immediately due to an active failing the LivenessFilter (as opposed to attempting the request anyway) or if a rebalance occurs, the active could be very behind and fail the MaximumLagFilter because even the active is very stale.  Of course these only happen if you enable heartbeats and lag reporting.  Otherwise they're noops.

But you're right that the primary use is to read from standbys.  If you don't think there's value in the above, I'll put them in an else clause.

Reply:
We should have the liveness filter enabled regardless..  for the same reason you cite. 

> They would allow for failing the request immediately due to an active failing the LivenessFilter
Repeatedly trying to open a tcp socket in a fast loop will kill latency. and also server CPU.

the ActiveHostFilter can be turned off ofc if standbys are not enabled.

Reply:
sounds good :)

Reply:
@vinothchandar Not sure what you mean by turning off the `ActiveHostFilter`. We need that to keep only the active (in the list of hosts to try) when routing to standbys is not enabled.